### PR TITLE
DM-21052: Update the pipe_analysis scripts to have the option to apply fgcm calibrations

### PR DESCRIPTION
**Pull request by laurenam**

### CODE REVIEW

File: python/lsst/pipe/analysis/analysis.py [hunk 1]
```diff
@@ -548,10 +548,13 @@ def plotAgainstMagAndHist(self, log, filename, stats=None, camera=None, ccdList=
 
         labelVisit(filename, plt, axScatter, 1.18, -0.11, color="green")
         if zpLabel is not None:
+            yOff = 0.02 if uberCalLabel is not None else 0
+            yOff = -0.02 if "_2" in zpLabel else yOff
```
Comment (erykoff) [inline python/lsst/pipe/analysis/analysis.py:555]: This logic is hard to follow.  I think it's worth splitting it apart into multiple explicit if/elif/else statements.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 2]
```diff
@@ -227,8 +227,8 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
         subdir = "patch-" + str(patchList[0]) if len(patchList) == 1 else subdir
         filenamer = Filenamer(repoInfo.butler, self.outputDataset, repoInfo.dataId, subdir=subdir)
         # Find a visit/ccd input so that you can check for meas_mosaic input (i.e. to set uberCalLabel)
-        self.uberCalLabel = determineUberCalLabel(repoInfo, patchList[0], coaddName=self.config.coaddName)
-        self.log.info("Uber-calibration used: {:}".format(self.uberCalLabel))
+        self.uberCalLabel = determineExternalCalLabel(repoInfo, patchList[0], coaddName=self.config.coaddName)
+        self.log.info("External calibration(s) used: {:}".format(self.uberCalLabel))
```
Comment (erykoff) [inline python/lsst/pipe/analysis/coaddAnalysis.py:232]: So there's no way of telling from the coadds what calibrations were applied, other than inferring from what exists in the repo?  I think this is a problem since we can mix and match types of calibrations.  If it's true that this sort of provenance information isn't available then we need to file a ticket.  (I just asked on slack.  I think this is important.)

File: python/lsst/pipe/analysis/utils.py [hunk 3]
```diff
@@ -1249,18 +1254,22 @@ def getRepoInfo(dataRef, coaddName=None, coaddDataset=None, doApplyUberCal=False
...
         coadd = butler.get(coaddName + coaddImageName, dataId)
         wcs = coadd.getWcs()
-        dataset = coaddName + coaddDataset
+        catDataset = coaddName + coaddDataset
         tractInfo = skymap[dataId["tract"]]
-    if doApplyUberCal:
-        dataset = "wcs_hsc" if hscRun else "jointcal_wcs"
+    photoCalibDataset = None
+    if doApplyExternalPhotoCalib:
+        photoCalibDataset = "fcr_hsc" if hscRun else externalPhotoCalibName + "_photoCalib"
```
Comment (erykoff) [inline python/lsst/pipe/analysis/utils.py:1269]: This is for meas_mosaic, right?  Is there a reason this isn't deprecated?  (Would it still run, even?) And, what does `hscRun` mean?  If this is meas_mosaic related, then maybe this should be renamed `measMosaicRun` or something?

File: python/lsst/pipe/analysis/visitAnalysis.py [hunk 4]
```diff
@@ -255,22 +273,61 @@ def runDataRef(self, dataRefList, tract=None, subdir=""):
...
-                                   "If not, run with --config doApplyUberCal=False".format(repoInfo.dataset))
-                raise RuntimeError("No datasets found for datasetType = {:s}".format(repoInfo.dataset))
+                raise RuntimeError("No datasets found for datasetType = {:s}".format(repoInfo.catDataset))
+            if self.config.doApplyExternalPhotoCalib:
+                ccdPhotoCalibListPerTract = getDataExistsRefList(dataRefListTract, repoInfo.photoCalibDataset)
+                if not ccdPhotoCalibListPerTract:
+                    self.log.fatal("No data found for {} dataset...are you sure you ran the external "
+                                   "photometric calibration?  If not, run with "
+                                   "--config doApplyExternalPhotoCalib=False".
+                                   format(repoInfo.photoCalibDataset))
```
Comment (erykoff) [inline python/lsst/pipe/analysis/visitAnalysis.py]: These and the below strings might be good candidates for the f-strings, e.g.:
```
self.log.fatal(f"No data found for {repoInfo.photoCalibDataset}...are you sure you ran the external "
               "photometric calibration?  If not, run with "
               "--config doApplyExternalPhotoCalib=False")
```

File: python/lsst/pipe/analysis/visitAnalysis.py [hunk 5]
```diff
@@ -716,19 +796,68 @@ def calibrateCatalogs(self, dataRef, catalog, fluxMag0, photoCalibDataset):
...
-    doApplyUberCal1 = Field(dtype=bool, default=True, doc="Apply uberCal (jointcal/meas_mosaic) results " +
-                            "to input1? FLUXMAG0 zeropoint from SFM is applied if doApplyUberCal is False")
+    doApplyExternalPhotoCalib1 = Field(dtype=bool, default=True,
+                                       doc=("Whether to apply external photometric calibration (e.g. "
+                                            "fgcmcal/jointcal/meas_mosaic) via an "
+                                            "`lsst.afw.image.PhotoCalib` object to input1.  If `True`, "
+                                            "uses ``externalPhotoCalibName1`` field to determine which "
+                                            "calibration to load.  If `False`, the instrumental flux "
+                                            "corresponding to 0th magnitude, fluxMag0, from SFM is "
+                                            "applied."))
```
Comment (erykoff) [inline python/lsst/pipe/analysis/visitAnalysis.py:816]: I was wondering if there was an easier way of doing this with less copy/pasting by making a sub-config and having it linked to twice, but it probably isn't worth it at this point.  But something to consider.

File: python/lsst/pipe/analysis/visitAnalysis.py [hunk 6]
```diff
@@ -1133,4 +1339,11 @@ def calibrateCatalogs(self, dataRef, catalog, fluxMag0, photoCalibDataset, doApp
             zpLabel = "FLUXMAG0"
             calibrated = calibrateSourceCatalog(catalog, self.zp)
 
+        if doApplyExternalSkyWcs:
+            wcs = dataRef.get(repoInfo.skyWcsDataset)
+            for record in catalog:
+                record.updateCoord(wcs)
```
Comment (erykoff) [inline python/lsst/pipe/analysis/visitAnalysis.py]: This can be done vectorized (and faster!) with `afwTable.updateSourceCoords(wcs, catalog)`, see e.g. https://github.com/lsst/validate_drp/blob/7546aab731893bfa50f0c0f189572c2f86b2591b/python/lsst/validate/drp/matchreduce.py#L318

**Review by erykoff — APPROVED** [review]
See comments below, and I trust the remainder "uber" references will be removed soon.  There are a couple of easy optimization I recommend.  I guess from Jim's answer on slack https://lsstc.slack.com/archives/C2JPMCF5X/p1582130598078300 that while you can get the external calibrations applied to the coadd, you can't actually get the name of them.  So the complicated repo-checking it is!

File: python/lsst/pipe/analysis/analysis.py (hunk 1 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/analysis.py:555]: I tried, but it actually didn't look much clearer.  I opted to leave it as is, but added the following explanatory comment:
>\# The following sets yOff to accommodate the longer labels for the
>\# compare scripts and/or for the presence of the extra uberCalLabel
>\# (for coadds).  Font size is adjusted with uberFontSize below.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 2 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py:232]: Agreed!  Sigh...

File: python/lsst/pipe/analysis/utils.py (hunk 3 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/utils.py:1269]: `meas_mosaic` is indeed effectively deprecated.  I haven't yet stripped out accommodations for `meas_mosaic` calibrations as runs still exist in the wild and "just in case" (there is still the occasional murmur about having to run it on the HSC calls...this would be very sad, but why spend time ripping out the functionality until all murmurs subside?).   As for `hscRun`, this actually means "was run using the HSC-stack", i.e. the forked version of many years ago.  This has long been retired, and ripping out this functionality is on the to-do list of https://jira.lsstcorp.org/browse/DM-14559 (but low-priority as it "does no harm").

>     hscRun : `str` or `None`, optional
>        String representing "HSCPIPE_VERSION" fits header if the data were
>        processed with the (now obsolete, but old reruns still exist)
>        "HSC stack".  Default is `None`.
>     matchRadius : `float` or `None`, optional
>

File: python/lsst/pipe/analysis/visitAnalysis.py (hunk 4 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/visitAnalysis.py]: I replaced most occurrences touched here, but will leave a full conversion (if that is what our standards would prefer) to a later ticket (likely DM-14559).

File: python/lsst/pipe/analysis/visitAnalysis.py (hunk 5 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/visitAnalysis.py:816]: I also consider this every time I make such updates...not having an immediate solution come to mind, I also concluded it's not worth it at this point.

File: python/lsst/pipe/analysis/visitAnalysis.py (hunk 6 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/visitAnalysis.py]: Awesome...I like faster!  Thanks for the tip 👍

**Review by laurenam — COMMENTED** [review]
Let me know if you're ok with things now.

**Comment by erykoff** [comment]
All good!
